### PR TITLE
add `cycle` method to LazyList

### PR DESCRIPTION
**Pull request by SethTisue**
I'm looking for multiple kinds of review here:

1) is this the right way, in this repo, to add an extension method to an existing type? we need to have at least one demonstration of what the "right" way is
2) do we, in fact, want to add this method? (I do genuinely want to add it, as the tie-the-knot trick is non-obvious, and I have seen people get it wrong, or not realize that it's possible; see https://stackoverflow.com/questions/2097851/scala-repeat-a-finite-list-infinitely and https://www.reddit.com/r/scala/comments/eih0t7/does_213s_lazylist_have_an_api_like_haskells_cycle/)
3) why does `testConstantMemory1` fail? @NthPortal is this as-expected?

if folks agree this should be added, I can do a better job on the Scaladoc

### CODE REVIEW

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (NthPortal) [inline src/main/scala/collection/immutable/LazyListExtensions.scala:27]: `result` doesn't need to be lazy I don't think

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (NthPortal) [inline src/main/scala/collection/immutable/LazyListExtensions.scala:24]: should check `knownSize == 0` (to emulate the internal `knownIsEmpty`)

File: src/test/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [inline src/test/scala/collection/immutable/TestLazyListExtensions.scala:66]: the vast majority of methods on `LazyList` (`drop` included) do not return cyclic `LazyList`s when called on cyclic `LazyList`s (see https://github.com/scala/bug/issues/11270#issuecomment-452515304 and https://github.com/scala/scala/pull/7692#issuecomment-458718024)

File: src/test/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [inline src/test/scala/collection/immutable/TestLazyListExtensions.scala:67]: this assertion is incorrect; rather, `cyc.state eq cyc.tail.tail.tail.state` (although `state` isn't public, so you can't test this easily)

**Review by NthPortal — COMMENTED** [review]
This does seem useful, though I don't like the fact that it requires you to construct a `LazyList` instance that doesn't get used

File: src/test/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [reply src/test/scala/collection/immutable/TestLazyListExtensions.scala:66]: actually, I take that back. I feel like `drop` *should* return a cyclic `LazyList`

**Comment by NthPortal** [comment]
@SethTisue do you mind if I push some changes?

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (NthPortal) [reply src/main/scala/collection/immutable/LazyListExtensions.scala:24]: actually, it's better to just wrap this in `LazyList.empty[T] #::: {` `}`

**Comment by SethTisue** [comment]
> do you mind if I push some changes?

@NthPortal have at it

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (NthPortal) [reply src/main/scala/collection/immutable/LazyListExtensions.scala:27]: apparently this is only the case when it's a member of a class/object

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (sjrd) [inline src/main/scala/collection/immutable/LazyListExtensions.scala:17]: With that name, we're going to need a separate `implicit class` for every extension that we add to `LazyList`. I would name this `LazyListOps` or `LazyListExtensionsOps`.

**Comment by SethTisue** [comment]
Luis (@BalmungSan) noticed that I forgot `extends AnyVal`

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (NthPortal) [reply src/main/scala/collection/immutable/LazyListExtensions.scala]: we also have to be mindful of the fact that in Scala 2, the name of that implicit class will shadow any other implicit class with that name, from any library

File: src/main/scala/scala/collection/immutable/LazyListExtensions.scala
Comment (julienrf) [inline src/main/scala/scala/collection/immutable/LazyListExtensions.scala:13]: Just curious: why is this `LazyList.empty` required?

Comment (NthPortal) [reply src/main/scala/scala/collection/immutable/LazyListExtensions.scala:13]: it makes the rhs lazy

Comment (SethTisue) [reply src/main/scala/scala/collection/immutable/LazyListExtensions.scala:13]: It's awkward that all the usual ways of constructing `LazyList`s require strict knowledge of whether the list is empty or not. `LazyList.empty #:::` is a workaround for what is arguably a gap in the API. We discussed it previously here: https://github.com/scala/bug/issues/11984#issuecomment-625528485

**Review by NthPortal — COMMENTED** [review]
1. I'm still thinking on this one. I think we might want it in a `next` sub-package?
2. I think it's quite valuable, and highly motivated given the links you provided.
3. The states were `eq` but not the `LazyList`s themselves; I've fixed the tests.

**Comment by SethTisue** [comment]
the Travis-CI .org to .com migration didn't happen smoothly here — separate ticket on that: https://github.com/scala/scala-library-next/issues/27

**Comment by SethTisue** [comment]
@NthPortal does the commit I just pushed line up with what you're proposing over at #24?

File: src/main/scala/collection/immutable/LazyListExtensions.scala
Comment (SethTisue) [reply src/main/scala/collection/immutable/LazyListExtensions.scala]: I went with `NextLazyListExtensions`, but I can change it again based on however the discussion at #24 turns out

File: src/main/scala/scala/collection/immutable/next/package.scala
Comment (NthPortal) [inline src/main/scala/scala/collection/immutable/next/package.scala:17]: should be `private val`

**Comment by NthPortal** [comment]
> does the commit I just pushed line up with what you're proposing over at #24?

yup

File: src/main/scala/scala/collection/immutable/next/package.scala
Comment (SethTisue) [reply src/main/scala/scala/collection/immutable/next/package.scala:17]: 😖, can’t wait for `extension`

**Comment by SethTisue** [comment]
@NthPortal additional commits pushed with more comments, more tests, more Scaladoc

one thing I discovered is that (even for finite `xs`) `xs.cycle.cycle` isn't itself a cycle — see line commented TODO. I'm thinking we probably don't care and it isn't worth adding a cyclical check to catch it? wdyt?

**Comment by NthPortal** [comment]
I don't know. I'm not sure what the cost of checking if it's a cycle is.

The more I look at this, the more I think cyclic `LazyList`s should be a separate type, but alas, it's a bit late for that. I think for now it's easiest not to worry about it? you could open a ticket and assign it to me and I'll see if I can figure it out at some point

**Comment by SethTisue** [comment]
> you could open a ticket

meh, I'm fine with it as it stands

**Comment by NthPortal** [comment]
sorry. I'll try to open a ticket or something. tomorrow. if I remember

**Comment by SethTisue** [comment]
> The more I look at this, the more I think cyclic LazyLists should be a separate type, but alas, it's a bit late for that

It depends on how deep one expects the support to go. For example, one might expect `.map` or `.filter` on a cycle to also return a cycle — they certainly could! But I don't think we want to go there.

It's me who (iirc) got us both thinking about cycles in the first place, before 2.13.0 was released — and honestly I was bit surprised that you ran with it and added the cycle-checking code to `LazyList` at all. I wouldn't say I regret that, but if we go one step further and add this `cycle` method, we probably need to limit user expectations in the Scaladoc for the new method. Like, do we document that `tail` and `drop` are expected to preserve cyclicality but other methods do not? Do we figure out _exactly_ which methods are cycle-preserving, or just issue a blanket disclaimer? Sigh.

**Comment by Ichoran** [comment]
@SethTisue - Actually, I think if we're going to support cycles at all, except as an accident, we should support them on `map` and `filter` and everything else.  It's really not fun to have a feature that randomly explodes on you because you "used it wrong", despite using it wrong looking like ordinary correct usages.

**Comment by NthPortal** [comment]
> honestly I was bit surprised that you ran with it and added the cycle-checking code to `LazyList` at all.

if I understand what you're referring to, I didn't end up doing that I don't think (other than what was already there in `force` and `addString`). what I added was a check for a self-referential (not necessarily cyclic) `LazyList` where you run out of elements but are still trying to derive something to append. I basically changed it from overflowing the stack to throwing a slightly nicer exception. unless I'm misunderstanding what you're talking about

**Comment by NthPortal** [comment]
> we should support them on `map` and `filter` and everything else.

I would love to, but I genuinely do not think it's possible without leaking a ton of memory and/or taking a huge performance hit, which is also undesirable. perhaps we could add cycle-aware variants of operations?

I *do* think we should add documentation to the stdlib's docs (and have the docs for this point to that?) that most operations do not preserve cycles, and perhaps the few that do.

> if we're going to support cycles at all, except as an accident

we have explicitly supported/handled cycles in `force` and `addString`/`toString` since the days of `Stream` (at least 2.11); we've just never documented or made an easy way to create them.

**Comment by SethTisue** [comment]
@NthPortal re: the history on this, I trust your memory over mine

> perhaps we could add cycle-aware variants of operations?

I don't want to work on that, and think we probably don't want to clutter `LazyList` up with that. I lean towards the just-document-which-methods-are-known-to-preserve-cycles approach

> Actually, I think if we're going to support cycles at all, except as an accident, we should support them on map and filter and everything else. It's really not fun to have a feature that randomly explodes on you because you "used it wrong", despite using it wrong looking like ordinary correct usages.

@Ichoran in practice, I don't think that's going to happen. I think our choices in practice are: add `cycle`, but document its limitations; or don't add it.

given that cycles are already "accidentally" supported, I feel like adding `cycle` is a good opportunity to: 1) make sure if that if someone wants a cycle they use our correct implementation instead of rolling their own, and 2) have an obvious place to put documentation about how cycles are handled. so I lean towards going ahead with the addition, but it's only a mild leaning.

**Comment by Ichoran** [comment]
@NthPortal - Yes, I added the cycle-handling.  It would be time-consuming but not, I think, all that computationally expensive to add it to most methods, as long as we don't guarantee zero unrolling of loops.  (If we did, we'd have to represent them explicitly.)  If we instead guarantee at most one unroll of a loop, we'd spend at most 1/2 an extra traversal on non-cyclic `LazyList`s.  Compared to the actual operation (map, etc.) which does a lot more work, that doesn't seem very expensive to me, especially on a collection that isn't really built for extreme performance to begin with.

**Comment by SethTisue** [comment]
added to Scaladoc:

```diff
+     *
+     * Note that some `LazyList` methods preserve cyclicality and others do not.
+     * So for example the `tail` of a cycle is still a cycle, but `map` and `filter`
+     * on a cycle do not return cycles.
```

I think I'm now satisfied with this.

File: src/main/scala/scala/collection/immutable/next/package.scala
Comment (NthPortal) [inline src/main/scala/scala/collection/immutable/next/package.scala:40]: this is not the case; if you allowed an empty `LazyList` through to case 3, this would happen:

```scala
scala> val ll = LazyList.empty[Int]
val ll: scala.collection.immutable.LazyList[Int] = LazyList()

scala> lazy val result: LazyList[Int] = ll #::: result
lazy val result: LazyList[Int] // unevaluated

scala> result.isEmpty
java.lang.RuntimeException: self-referential LazyList or a derivation thereof has no more elements
  at scala.collection.immutable.LazyList.scala$collection$immutable$LazyList$$state$lzycompute(LazyList.scala:256)
  at scala.collection.immutable.LazyList.scala$collection$immutable$LazyList$$state(LazyList.scala:252)
  at scala.collection.immutable.LazyList.$anonfun$lazyAppendedAll$1(LazyList.scala:379)
  at scala.collection.immutable.LazyList.scala$collection$immutable$LazyList$$state$lzycompute(LazyList.scala:259)
  at scala.collection.immutable.LazyList.scala$collection$immutable$LazyList$$state(LazyList.scala:252)
  at scala.collection.immutable.LazyList.isEmpty(LazyList.scala:269)
  ... 32 elided
```

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [inline src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:24]: is there something in sbt that will stop this if it's wrong? I'm not super keen on the fail state of a test being hanging

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [inline src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:112]: I don't think we care. If we do, I'm not sure how we'd do it. _technically_ it is currently correct (the worst kind of correct), as the first `.cycle` returns an infinite `LazyList`, so the cycle length of the second `.cycle` should be infinite

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [inline src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:38]: I have bad news for you... (see [LazyList.scala#L1156](https://github.com/scala/scala/blob/v2.13.4/src/library/scala/collection/immutable/LazyList.scala#L1156))

```scala
scala> LazyList().knownSize
val res1: Int = 0
```

you might try `LazyList(1).drop(1)`?

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (NthPortal) [inline src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:29]: I think these are good tests, but if they are all for `cycle`, which is not indicated anywhere in the method names. if we ever add another extension method, the test names will be a bit confusing. perhaps either `emptyCycle1` or `cycleEmpty1`.

also, small nit: I don't think the methods need to have `test` in their names.

**Review by NthPortal — COMMENTED** [review]
looks pretty good to me; I think it needs just a few (mostly cosmetic) changes

File: src/main/scala/scala/collection/immutable/next/package.scala
Comment (NthPortal) [reply src/main/scala/scala/collection/immutable/next/package.scala:40]: it would be kind of as if you tried to have an infinitely long `ll #::: ll #::: ll #::: ll #::: ll #::: // ...`

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (SethTisue) [reply src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:24]: well, I'm not keen on it either... I'll add this comment:

```
  // (It's kind of nasty to have tests whose failure mode is to hang, but I don't                                        
  // see an obvious alternative that doesn't involve copying code from LazyList.                                        
  // Perhaps this could be improved at the time this all gets merged into stdlib.)                                      
```

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (SethTisue) [reply src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:112]: comment updated.

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (SethTisue) [reply src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:38]: I'll go with `LazyList.empty #::: LazyList.empty` and add a `knownSize` check

File: src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala
Comment (SethTisue) [reply src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala:29]: I took both of these suggestions.

**Comment by SethTisue** [comment]
review feedback, handled

**Review by NthPortal — APPROVED** [review]
LGTM! ty Seth!

**Comment by SethTisue** [comment]
not for merge yet...

...since I'm not sure yet if we have sufficient consensus yet on actually adding this — I don't really know who watches this repo and who doesn't. so let's hold off on merging — I think we'll need to institute some (not-TOO-overly-formal) process for that involving signoff from both the Scala 2 team at Lightbend and the Scala 3 team at EPFL

**Comment by SethTisue** [comment]
Well this has now been sitting in limbo for a whole year since no one (including me) has stepped forward to formulate an approval process.

I'm going to close it to get it out of my personal PR list. We can find it again with the "status:pending" label.

**Pull request closed without merging**
